Hometask 1

kkonanov opened the pull request.

Commit:
[task1] change Application

Commit:
[task1] change Application

a-ivanov commented on pom.xml:
```diff
@@ -25,12 +25,20 @@
             <scope>test</scope>
         </dependency>
+
+<!--        <dependency>-->
```
Удалите закомменченный код.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SmartHomeReader.java:
```diff
@@ -0,0 +1,20 @@
+import java.nio.file.Paths;
+
+public class SmartHomeReader {
+    public SmartHome readFromJs(String filename) {
```
Добавьте интерфейс, это абстракция чтения состояния дома.

a-ivanov replied:
В названии метода интерфейса, конечно же не дожно быть деталей тиипа JS.
Еще помните про https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#%D0%BD%D0%B5-%D0%B4%D0%B0%D0%B2%D0%B0%D0%B9%D1%82%D0%B5-%D0%B0%D0%B1%D1%81%D1%82%D1%80%D0%B0%D0%BA%D1%86%D0%B8%D0%B8-%D1%82%D0%B5%D1%87%D1%8C

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Application.java:
```diff
@@ -9,72 +9,15 @@
-        return new SensorEvent(sensorEventType, objectId);
+    public static void main(String... args) {
+        Application application = new Application();
+        application.sensorEventCicle.run();
```
Прячьте детали application.sensorEventCicle... -> application.run(), метод main не должен знать детали, как приложение работает.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SensorEventGenerator.java:
```diff
@@ -0,0 +1,11 @@
+
+public class SensorEventGenerator {
+
+	public static SensorEvent getNextSensorEvent() {
```
https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#%D0%BD%D0%B5-%D0%B8%D1%81%D0%BF%D0%BE%D0%BB%D1%8C%D0%B7%D1%83%D0%B9%D1%82%D0%B5-%D1%81%D1%82%D0%B0%D1%82%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%B8%D0%B5-%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D1%8B-%D0%B2-%D0%B1%D0%B8%D0%B7%D0%BD%D0%B5%D1%81-%D0%BB%D0%BE%D0%B3%D0%B8%D0%BA%D0%B5

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SensorEventCicle.java:
```diff
@@ -0,0 +1,18 @@
+		this.smartHome = smartHome;
+	}
+	void run() {
+		SensorEventHandler sensorEventHandler = new SensorEventHandler(smartHome);
```
https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#%D0%BF%D0%B5%D1%80%D0%B5%D0%B4%D0%B0%D0%B2%D0%B0%D0%B9%D1%82%D0%B5-%D0%B7%D0%B0%D0%B2%D0%B8%D1%81%D0%B8%D0%BC%D0%BE%D1%81%D1%82%D0%B8-%D1%81%D0%BD%D0%B0%D1%80%D1%83%D0%B6%D0%B8

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SensorEventHandler.java:
```diff
@@ -0,0 +1,22 @@
+
+    public void handle(SensorEvent sensorEvent) {
+        System.out.println("Got event: " + sensorEvent);
+        if (sensorEvent.getType() == LIGHT_ON || sensorEvent.getType() == LIGHT_OFF) {
```
https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#%D0%BF%D0%B5%D1%80%D0%B5%D0%B4%D0%B0%D0%B2%D0%B0%D0%B9%D1%82%D0%B5-%D0%B7%D0%B0%D0%B2%D0%B8%D1%81%D0%B8%D0%BC%D0%BE%D1%81%D1%82%D0%B8-%D1%81%D0%BD%D0%B0%D1%80%D1%83%D0%B6%D0%B8

a-ivanov replied:
Нарушение OCP: добавление нового обработчика требудет модификацию кода.
Выделите интерфейс обработчика, имплементируйте для каждого вида событий. Перенесите туда логику запуска обработки для конкретного события.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/DoorEventHandler.java:
```diff
@@ -0,0 +1,33 @@
+						System.out.println("Door " + door.getId() + " in room " + room.getName() + " was closed.");
+						// если мы получили событие о закрытие двери в холле - это значит, что была закрыта входная дверь.
+						// в этом случае мы хотим автоматически выключить свет во всем доме (это же умный дом!)
+						if (room.getName().equals("hall")) {
```
https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#%D1%80%D0%B0%D0%B7%D0%B4%D0%B5%D0%BB%D1%8F%D0%B9%D1%82%D0%B5-%D0%B8%D0%B7%D0%BC%D0%B5%D0%BD%D0%B5%D0%BD%D0%B8%D0%B5-%D0%BE%D0%B1%D1%8A%D0%B5%D0%BA%D1%82%D0%BE%D0%B2-%D0%B8-%D1%81%D1%86%D0%B5%D0%BD%D0%B0%D1%80%D0%B8%D0%B8-%D1%83%D0%BC%D0%BD%D0%BE%D0%B3%D0%BE-%D0%B4%D0%BE%D0%BC%D0%B0

Commit:
[task1]
move classes to folders
add interface SmartHomeReader and implementation SmartHomeReaderFromJs
add method Application.run()
add interface EventGenerator and implementation EventGeneratorImpl
add interface EventHandler and implementations

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Application.java:
```diff
@@ -1,80 +1,27 @@
-            event = getNextSensorEvent();
-        }
+    Application(SmartHomeReader smartHomeReader, EventGenerator eventGenerator) {
+        smartHome = smartHomeReader.read();
```
Констурктор - только для присвоения аргументов полям класса. Все действия (чтение, расчеты, итд) лучше делать в методах класса.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Application.java:
```diff
@@ -1,80 +1,27 @@
-        }
+    Application(SmartHomeReader smartHomeReader, EventGenerator eventGenerator) {
+        smartHome = smartHomeReader.read();
+        eventCircle = new EventCircleImpl(smartHome, eventGenerator);
```
https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#%D0%BF%D0%B5%D1%80%D0%B5%D0%B4%D0%B0%D0%B2%D0%B0%D0%B9%D1%82%D0%B5-%D0%B7%D0%B0%D0%B2%D0%B8%D1%81%D0%B8%D0%BC%D0%BE%D1%81%D1%82%D0%B8-%D1%81%D0%BD%D0%B0%D1%80%D1%83%D0%B6%D0%B8

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/eventhandler/DoorEventHandler.java:
```diff
@@ -0,0 +1,62 @@
+	private void doorClosed(Room room, Door door) {
+		door.setOpen(false);
+		System.out.println("Door " + door.getId() + " in room " + room.getName() + " was closed.");
+		if (room.getName().equals("hall")) {
```
https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#%D1%80%D0%B0%D0%B7%D0%B4%D0%B5%D0%BB%D1%8F%D0%B9%D1%82%D0%B5-%D0%B8%D0%B7%D0%BC%D0%B5%D0%BD%D0%B5%D0%BD%D0%B8%D0%B5-%D0%BE%D0%B1%D1%8A%D0%B5%D0%BA%D1%82%D0%BE%D0%B2-%D0%B8-%D1%81%D1%86%D0%B5%D0%BD%D0%B0%D1%80%D0%B8%D0%B8-%D1%83%D0%BC%D0%BD%D0%BE%D0%B3%D0%BE-%D0%B4%D0%BE%D0%BC%D0%B0

Commit:
[task1]
corrected

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/eventcircle/EventCircleImpl.java:
```diff
@@ -0,0 +1,35 @@
+	public EventCircleImpl(SmartHome smartHome, EventGenerator eventGenerator) {
+		this.smartHome = smartHome;
+		this.eventGenerator = eventGenerator;
+		eventHandlers = Arrays.asList(new DoorEventHandler(smartHome), new LightEventHandler(smartHome), new AllLightEventHandler(smartHome));
```
https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#%D0%BF%D0%B5%D1%80%D0%B5%D0%B4%D0%B0%D0%B2%D0%B0%D0%B9%D1%82%D0%B5-%D0%B7%D0%B0%D0%B2%D0%B8%D1%81%D0%B8%D0%BC%D0%BE%D1%81%D1%82%D0%B8-%D1%81%D0%BD%D0%B0%D1%80%D1%83%D0%B6%D0%B8
